src/components/checkbox/GroupableCheckboxReducers.ts

```diff
@@ -0,0 +1,171 @@
+import * as _ from 'underscore';
+import { IReduxAction } from '../../utils/ReduxUtils';
+import { ICheckboxState } from './CheckboxReducers';
+import { GroupableCheckboxActions, IGroupableCheckboxActionPayload } from './GroupableCheckboxActions';
+
+export interface IGroupableCheckboxesState {
+  total: number;
+  nbSelected: number;
+  parentId?: string;
+  parent?: ICheckboxState;
+  checkboxes: ICheckboxState[];
+}
+
+export const groupableCheckboxInitialState: IGroupableCheckboxesState = {total: 0, nbSelected: 0, checkboxes: []};
+export const groupableCheckboxesInitialState: IGroupableCheckboxesState[] = [];
+
+export const addParentCheckbox = (state: IGroupableCheckboxesState,
+                                  action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  const newCheckbox = {
+    id: action.payload.id,
+    checked: !!action.payload.checked,
+  };
+
+  if (state === groupableCheckboxInitialState) {
+    return {
+      ...state,
+      parentId: action.payload.id,
+      parent: newCheckbox,
+    };
+  }
+
+  state.parent = newCheckbox;
+  return state;
+};
+
+export const addChildCheckbox = (state: IGroupableCheckboxesState,
+                                 action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  const newChecboxes: ICheckboxState[] = state.checkboxes.concat({
+    id: action.payload.id,
+    checked: !!action.payload.checked,
+  });
+
+  const total: number = state.total + 1;
+  const nbSelected: number = state.nbSelected + (action.payload.checked ? 1 : 0);
+
+  if (state === groupableCheckboxInitialState) {
+    return {
+      ...state,
+      total,
+      nbSelected,
+      checkboxes: newChecboxes,
+      parentId: action.payload.parentId,
+    };
+  }
+
+  return {...state,
+    checkboxes: newChecboxes,
+    total,
+    nbSelected,
+  };
```

Comment:
I'd replace this with:
```
const newState = {
  ...state,
  checkboxes: [
    ...state.checkboxes,
    newCheckbox,
  ],
  total: state.total + 1,
  nbSelected: state.nbSelected + (action.payload.checked ? 1 : 0),
};

if (state === groupableCheckboxInitialState) {
  newState. parentId = action.payload.parentId;
}

return newState
```

Less repetition and you aren't reassigning the state.